platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/LifecycleTrackingTestEventReporter.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.tasks.testing.TestEventReporter;
+import org.gradle.api.tasks.testing.TestOutputEvent;
+
+import java.time.Instant;
+
+@NonNullApi
+class LifecycleTrackingTestEventReporter<T extends TestEventReporter> implements TestEventReporter {
+    protected final T delegate;
```

Comment:
🤔 Considering there seems like there would only ever be one type to delegate to used here and in `StateTrackingTestEventReporter`, would it be better to use inheritance instead of delegation?

Then we could:
- Remove the need for 2 instances of `TestEventReporter` whenever we want to use one.
- Remove the need for a `DefaultTestEventReporter` which not really an actual default choice anywhere.  It could be renamed to `AbstractTestEventReporter` and concretized by `LifecycleTracking` and `StateTracking` variants. 
- Remove the need to reimplement methods in `StateTrackingGroupTestEventReporter` to just call `delegate.method()`, which would be simpler and less code.

Reply:
Because of `GroupTestEventReporter`, there are always two slightly different implementations. We have groups and tests, so `StateTrackingGroupTestEventReporter` and `StateTrackingTestEventReporter` have to do slightly different things.

e.g.,
```
Root 
  Group A
    Test 1
  Group B
    Test 1
```
Each of these levels will have a start/completed event, but we only want to count the leaves as the actual tests. If we treat groups the same as tests, this hierarchy will have 5 tests instead of 2 tests.

This does raise some question about how do we render failures at the group level that are not also failures at the test level (like Before/AfterClass methods).

I took another stab at pushing these back together and eliminated all delegation except between LifecycleTracking and the Default implementations.  I think this is better.

Reply:
I wasn't suggesting reworking the group vs. non group implementations, but the `StateTrackingTestEventReporter` and `LifecycleTrackingTestEventReporter` implementations.  So that instead of:

```
GroupTestEventReporter (interface)
├─ DefaultGroupTestEventReporter
│  ├─ LifecycleTrackingGroupTestEventReporter
```

and 2 instances of an event reporter at runtime (the `LifecycleTracking` one, and the delegate it contains), there would be:

```  
GroupTestEventReporter (interface)
├─ AbstractGroupTestEventReporter
│  ├─ LifecycleTrackingGroupTestEventReporter
```

and a single instance at runtime.